### DISCUSSION TITLE
Per-project display setup instead of editing always User_setup.h

### DISCUSSION THREAD

**Discussion by homonto** [Bodmer/TFT_eSPI#1689] [General]
Hi, 
I know I can uncomment 1 line and comment the other one in the "User_Setup_Select.h" file, to have my setup included (per display).
But, I am thinking about more robust way.
Why?
Since I have few displays and projects, it happened to me already few times to compile the project with "wrong" display.
Outcome? Nothing on display (of course).
Is there any possibility to do something like this in my ino file:
#define MY_DISPLAY_1
and then, in the file "User_Setup_Select.h" put something like this:
#ifdef MY_DISPLAY_1
#indclude <MY_DISPLAY_1.h>
#endif

I actually tried it but apparently it is not working - complaining too much - apparently such directive is messing around

If such mechanism exists, I would never ever mess up with my projects.

Thank you

**Comment by Bodmer** [comment]
The Arduino IDE does not allow sketch #defines to be visible to a library  hence the use of setup files.

If you use PlatformIO as the IDE then you can have project based setups in the platformio.ini files  see tools folder.

One possibility with the Arduino IDE is to generate a board type specific to a project but there are no tools to do this, and it would still be necessary to select the right board in the IDE when a different setup is wanted.

It might be possible to write an add in for the IDE that behaves as as project manager but I have not tried that.

Another option is to allow a sketch to detect the setup in use and generate a compile time error if it is wrong. That would be an easy to add feature.

**Comment by homonto** [comment]
thank you Bodmer - it seems Arduino is for lamers so I am one of them and I need to stay there as Platformio has other things that are not working i.e. S2 is still not there

**Reply by memeticae** [reply]
Only "lamers" think #define works outside of compile time.
(Remember a #define is only a macro.)

You will always need to know for which hardware you are compiling.
But you can always use something like visual basic, that works on all supported platforms, no nasty hardware questions for you.

;)

What would be REALLY useful is something like tft.queryspecs(interface)
But I know I'm long dead before something like that works flawlessly.

**Comment by homonto** [comment]
btw I made it working by using some macros - in case someone is interested in how (maybe dirty code but... working).
when I use ST7789 display I define in my ino file:
`#define ST7789_LCD`

for ILI I use:
`#define ILI9341_LCD`

I actually at this moment of time use only these two.

Then in  void setup():
```
  #if defined(ILI9341_DRIVER) and defined(ST7789_DRIVER)
//the above comes from your setup files where you define which driver is in use
    #error "2 LCD drivers defined - choose only ONE please"
    #error "in file User_Setup_Select.h "
  #endif

  #ifdef ILI9341_LCD
    #ifndef ILI9341_DRIVER
      #error "wrong display configured - change the line in User_Setup_Select.h file accordingly"
    #endif
  #endif

#ifdef ST7789_LCD
  #ifndef ST7789_DRIVER
    #error "wrong display configured - change the line in User_Setup_Select.h file accordingly"
  #endif
#endif

```

this way the compilation stops when wrong file is selected in User_Setup_Select.h
so thank you for the hints

**Comment by Bodmer** [comment]
I will be adding the capability to read info defined in the user_setup file at the next release.

Although the Arduino IDE prevents a #define in the sketch from being recognised by the library, the reverse is possible a library #define is visible to the sketch. So if you add things to the setup file like:
```
#define USER_SETUP_INFO "ST7796_Pico.h"
#define USER_SETUP_ID 1234
```
The sketch has access to these, so you can generate an error message is the wrong setup is being used. Add lines like this to the sketch:
```
#include <TFT_eSPI.h>      //  <<<< Must be before user ID check
#if defined (USER_SETUP_ID)
  #if (USER_SETUP_ID != 1234)
    #error WRONG setup file being used!
  #endif
#endif
```
Note that the #define is not available to the sketch until after the #include <TFT_eSPI.h>

**Comment by Bodmer** [comment]
I see you have already adopted the above approach. I am also adding an ID check function although that is strictly not necessary.
For example:
```
  if ( !tft.verifySetupID(1234) ) {
    while(1) {  // Loop here forever
      Serial.println("Wrong setup ID, please check!");
      delay(1000);
    }
  }
```

**Comment by homonto** [comment]
I am not sure if this is me only but whenever you update your library ALL FILES are overwritten with the new ones, right?
At least this is what I noticed.
For that reason my pins and some extras are defined in my user file, out of your library.
This way, whenever update happens, I only change 2 lines in User_Setup_Select, i.e. today it is like this:

```
// Only ONE line below should be uncommented.  Add extra lines and files as needed.

// #include <User_Setup.h>           // Default setup is root library folder


// Setup file configured for my ST7789 240x240
#include <../0-ZH_library/TFT_eSPI_display_setups/ZH_ST7789.h>                

// Setup file configured for my ILI9341 320x240 - both touch and NO touch - the same pins and settings
// #include <../0-ZH_library/TFT_eSPI_display_setups/ZH_ILI9341.h>  
```

and then I only uncomment 1st and comment 2nd

**Reply by Bodmer** [reply]
Yes,that is correct. The approach I use is documented [here](https://github.com/Bodmer/TFT_eSPI#tips). I have not come up with a better method unfortunatley :-(

**Comment by homonto** [comment]
I am sure I took it from there
But thx for confirmation that ... I am doing as needed ;-)
But, 1 question: many libraries define macros and if you don't redefine it, the default value is taken, but if you define in your sketch it will take the new one. So I would think like this:
Proposal 1:
1- in your library i.e.
#define MISO_GPIO 5
#define TFT_HEIGTH 240
etc
2- then if I am ok with that, I do nothing
3- if I want to redefine I do it in setup:
#define MISO_GPIO 6
#define TFT_HEIGTH 320
etc

wouldn't that work?

Proposal 2:
1- you expose the initialisation instead of:
tft.init();
2- to be:
tft.init(MISO_GPIO, MOSI_GPIO, CLK_GPIO, HEIGHT...) ... etc.
In that case GPIO definition could have been overwritten

I see i.e. that it can be either:
WiFi.begin(ssid, password);
or:
WiFi.begin(ssid, password, WIFI_CHANNEL);

I am thinking logically even? I am not a professional programmer ;-)

**Reply by Bodmer** [reply]
I think proposal 1 will not work as the Arduino IDE keeps the macros separate.

Yes, pins could be defined in the init or when the class instance is created. The problem there is that the compiler cannot use some of the performance optimisations. When I created the library I went around the "how to do this" loop a few times. In the end I decided that there were so many other config options that a setup file was the only answer for my needs. The great thing for me is that I can run any example on hardware setups without a single change to the example sketch. With Adafruit_GFX I used to forget to change the IO pins in a sketch when I swapped hardware projects, so you can't win sometimes!

It is a pity the Arduino IDE does not have an approach like PlatformIO, maybe the latest Arduino 2.0 RC IDE's will add that as I see they are adding debug capability.

You can put the whole TFT_eSPI library with a bespoke setup into each sketch but that then creates a new upgrade path problem.

**Reply by homonto** [reply]
ok, fair enough, and THANKS for great job!

but when I did not know you have defined color, i.e. TFT_BLACK and I defined it in the sketch, I got only warning "redefined" but it was still working - not sure which definition was taken - probably the last one ;-)

**Comment by moreauftheobald** [comment]
Could it bé possible to define insta thé sketch thé pin (formation example (#DEFINE TFT_MOSI 15)  instead offre into thé setup filé ? 
That would bé à great progress

**Reply by Bodmer** [reply]
The Arduino IDE prevents #defines in the sketch being visible to the library.

PlatformIO IDE does allow per project configuration, so is a good option if that is essential.

**Comment by Bodmer** [comment]
See here for a work-around that allows the setup file to be included in the sketch:

https://github.com/Bodmer/TFT_eSPI/discussions/2007#discussioncomment-3614189
